### ISSUE TITLE
Something similar in tauri?

### ISSUE THREAD

**Issue by integralfunction** [andrewbaxter/wongus#4]
I like the idea behind this project. From what I see wongus goes a little lower level and creates a new gtk window, does layer shell magic with gtk-layer-shell, then creates a wry context with this gtk. I'm looking to have backend and frontend communicate with each other and use something like react (not served from a server). Would it be possible to do something similar using tauri? Perhaps tauri exposing the gtk window somehow?

**Comment by andrewbaxter** [comment]
Are you working on making something similar?  Your reading of the code is accurate.

I can't say I'm an expert in Tauri, but my sense is that it basically wraps up a bunch of basic stuff (like static file serving for the frontent, wasm generation, and generating structures/shims for ipc calls on both sides).

I feel like Tauri must have an example with react because I think a lot of people would want to use that for the frontend, but with wry/tao directly you could probably do something similar to what I'm doing: you'd need to pack your javascript/html to get a static dir, add a custom protocol handler (like the code here does) to serve the static files, then set the default url to be a url with your custom protocol for `index.html`.

Is that what you were wondering about?

**Comment by integralfunction** [comment]
Yes, I'm trying to make something similar, so this isn't exactly a wongus issue, sorry if it's a bit unrelated. The main reason I want to use tauri is to have the backend and frontend communicate with each other. I will have to look into that more, thanks.

**Comment by andrewbaxter** [comment]
Oh no problem! Yeah, I think Tauri would probably be the most straightforward.  Serving react and running the backend in the same program shouldn't be a problem.

**Comment by LucaCoduriV** [comment]
I am looking to make exactly the same haha.
Funny to see that I am not alone investigating.

Last time I tried it was not possible to use tauri with layershell because I need the window handler before it was mapped.
@andrewbaxter  Do you have an Idea for this ?

@integralfunction Have you found something ?

I am thinking about forking tauri and try to change the builder struct

**Comment by integralfunction** [comment]
> Funny to see that I am not alone investigating

Yeah this is a really niche use case lol

@LucaCoduriV I haven't really looked much into tauri after that

For the window mapping issue you're referencing, I think there's a similar issue coincidentally made by @andrewbaxter :
https://github.com/tauri-apps/wry/issues/1253
If you follow it around, it leads to this pr being merged:
https://github.com/tauri-apps/tao/pull/938
which is the function used in wongus to make layer shell possible (`new_from_gtk_window()`)

However if I'm not wrong the [setup](https://docs.rs/tauri/1.7.2/tauri/struct.Builder.html#method.setup) method in tauri already contains an app with the gtk window hooked onto it, so I'm not sure if it is possible to use it in tauri (?)

I've looked into other options like the Chromium Embedded Framework but it's rust bindings seem unmaintained. I'd be happy to hear if you have any successes with layer shell in tauri.

**Comment by LucaCoduriV** [comment]
Yeah I think it's not possible sadly.
But I might have good news, it might be possible with [Dioxus](https://dioxuslabs.com/learn/0.5/reference/desktop).
I am going to try and tell you if I was able to make it work.

**Comment by LucaCoduriV** [comment]
I wasn't able to make it work, but I feel that it should be possible.
Dioxus codebase look way more simpler than Tauri.
If you are motivated to help it would be with pleasure.

Dioxus is based on TAO for windowing, and lately @andrewbaxter made a pr to add the from_gtk_window method.
Thanks to @andrewbaxter it should be very easy to create the gtk_window manually and use it with dioxus webview.

**Comment by integralfunction** [comment]
@LucaCoduriV

Hmmm, as I see dioxus uses the WindowBuilder which should be possible but the gtk window is created using the event loop. If there would be a way to expose the event loop then it should be possible. Do you know how?

```rs

fn main() {
  // how to get the dioxus event loop?
  let gtkwindow = gtk::ApplicationWindow::new(event_loop.gtk_app());
  // do layer shell stuff on gtk window
  let config = dioxus_desktop::Config::new()
    .with_custom_head(r#"<link rel="stylesheet" href="tailwind.css">"#.to_string())
    .with_window(
      Window::new_from_gtk_window(event_loop, gtkwindow).unwrap_or_else(|_error: OsError| {
        panic!("Could not create tao window from gtk window");
      });
  );
  LaunchBuilder::desktop().with_cfg(config).launch(App);
}
```

**Comment by LucaCoduriV** [comment]
Sadly it's going to be harder than that.
It's a bit weird but when you see window it means windowBuilder in dioxus.
So we need to modify the dioxus code to be able to pass a TAO::Window

I was thinking about a callback that takes the event loop as a parameter and in which we create the gtk window.

```rust
LaunchBuilder::desktop().from_tao_window(|event_loop|{

let gtkwindow = gtk::ApplicationWindow::new(event_loop.gtk_app()).unwrap();
  // do layer shell stuff on gtk window
 
gtkwindow

}).launch(App);
```

**Comment by LucaCoduriV** [comment]
Okay so I tried something you should be able to see the diff with this link: https://github.com/DioxusLabs/dioxus/compare/main...LucaCoduriV:dioxus:clean?expand=1

But I have an error because of gtk when I try to create the EventLoop
```
thread 'main' panicked at /home/luca/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tao-0.29.1/src/platform_impl/linux/event_loop.rs:189:53:
Failed to initialize gtk backend!: Error { domain: g-io-error-quark, code: 2, message: "An object is already exported for the interface org.gtk.Application at /org/gtk/Application/anonymous" }
```

Do you have an idea of why this is happening ?

I am so close to make this thing work ;(


Edit:
It works !!!! The error was due to me creating the Eventloop twice...
Here is my code if you want to check it: https://github.com/LucaCoduriV/dioxus/tree/clean

So now I need to clean my code and make it possible to create the window from the user side.
I'll come back to you when it's done.

**Comment by integralfunction** [comment]
Wow you made a solution fast, nice! Are you planning to make a pull request to merge it to the main repo?

**Comment by LucaCoduriV** [comment]
Thanks haha :)
I am not sure that they are going to accept it, at least not in the current state.
To not brake everything without modifying a lot of code I did some copy paste.

If you want to help me to think of an API that fits nice with the current code your are welcome.

I was thinking about adding a feature flag that adds layershell. What do you think ?
And then I would make a LayershellBuilder to build that layer.
What I am afraid of making a Builder is that it's going to be a pain in the ass to write all methods...

**Comment by integralfunction** [comment]
> I was thinking about adding a feature flag that adds layershell. What do you think ?

Yea sure that can be good, but why do you need a builder? Will it work if the feature just introduces the `_from_window()` functions that you wrote already?

**Comment by LucaCoduriV** [comment]
my initial idea was to take a callback to create the window, something like
`_from_window(|event_loop|{ //Create you window}|)`
But by doing so, I need to expose the event_loop of type UserWindowEvent and if the plan is to make a pr I am not sure that they are going to accept this.

So my second idea is to only expose a LayerShellBuilder that would exist only if the feature flag is on.

**Comment by andrewbaxter** [comment]
It might be a good idea to make an issue there first and ask them for design suggestions, it could make it easier to get merged.  If they do want it but want design changes, it can be hard to ask that after the PR is made.

**Comment by LucaCoduriV** [comment]
> It might be a good idea to make an issue there first and ask them for design suggestions, it could make it easier to get merged.  If they do want it but want design changes, it can be hard to ask that after the PR is made.

You're right, I'll try to ask them. But I am afraid that this will take too much time.

**Comment by andrewbaxter** [comment]
Oh yeah, true. Well TBH sometimes I start working on a patch before getting an answer in the issue too... :sweat_smile:

**Comment by LucaCoduriV** [comment]
Okay I made somethings usable.
If you use my fork you should be able to do something like this:

```rust
fn main() {
    // Init logger
    dioxus_logger::init(Level::INFO).expect("failed to init logger");
    info!("starting app");


    LaunchBuilder::desktop_from_gtk_window(
        |event_loop| {
            let gtk_window = gtk::ApplicationWindow::new(event_loop.gtk_app());
            gtk_window.set_app_paintable(true);
            gtk_window.set_decorated(false);
            gtk_window.stick();
            gtk_window.set_title("This is a wongus");
            let default_vbox = gtk::Box::new(gtk::Orientation::Vertical, 0);
            gtk_window.add(&default_vbox);

            gtk_window.init_layer_shell();
            gtk_window.set_layer(gtk_layer_shell::Layer::Top);
            gtk_window.auto_exclusive_zone_enable();
            gtk_window.set_anchor(gtk_layer_shell::Edge::Top, true);
            gtk_window.set_anchor(gtk_layer_shell::Edge::Right, true);
            gtk_window.set_anchor(gtk_layer_shell::Edge::Bottom, false);
            gtk_window.set_anchor(gtk_layer_shell::Edge::Left, false);
            gtk_window.set_width_request(400);
            gtk_window.set_height_request(400);
            gtk_window.show_all();
            let window =
                tao::window::Window::new_from_gtk_window(event_loop, gtk_window.clone())
                    .unwrap();
            // I have no idea if creating an Rc is needed here, but just in case added one
            (window, Rc::new(default_vbox))
        }
    ).launch(App);
}
```
The code behind isn't perfect at all, I just made a naive solution.
To make something clean I would need to change a lot of code that they've written.

I still didn't made an issue about this, I might do it tomorrow.

But in the meanwhile it works :)
https://github.com/LucaCoduriV/dioxus/tree/from_gtk_window

**Comment by MrAdhit** [comment]
Hey, after I did some hacking on the Tauri codebase, I was able to get layer shell working.

But I encountered this issue where if I set the background to transparent, it left some kind of ghosting from the previous frame, the "ghost" persist for some time and it cleared itself.

Do any of you know why?

![2024-10-12-033125_hyprshot](https://github.com/user-attachments/assets/66d5ee7b-74a7-4dc9-8a1d-dea689edcc92)

For more context, I'm able to expose the `gtk_app` and use it to create a new gtk window, and move the existing vbox to that window
```rs
fn setup(app: &mut App) -> anyhow::Result<()> {
    let webview_window = app.get_webview_window("Main").unwrap();
    webview_window.hide().unwrap();
    webview_window.set_decorations(false).unwrap();
    
    let target_monitor = gtk::gdk::Display::default().unwrap()
        .monitor(0).unwrap();
    
    let window = gtk::ApplicationWindow::new(webview_window.gtk_app());
    
    window.set_app_paintable(true);
    window.set_decorated(false);
    window.stick();
    
    webview_window.gtk_window().unwrap().remove(&webview_window.default_vbox().unwrap());
    window.add(&webview_window.default_vbox().unwrap());

    window.init_layer_shell();

    window.auto_exclusive_zone_enable();

    window.set_layer(Layer::Top);
    window.set_anchor(Edge::Left, true);
    window.set_width_request(50);
    window.set_height_request(target_monitor.geometry().height());

    window.show_all();

    window.set_monitor(&target_monitor);

    Ok(())
}
```

**Comment by MrAdhit** [comment]
Btw, idk if I'm dumb or what 😅
I don't need to hack tauri to get the gtk app, I was able to get it working by just using 
`&webview_window.gtk_window().unwrap().application().unwrap()`

And yes, the ghosting issue still persist, and I still need help with that

```rs
fn setup(app: &mut App) -> anyhow::Result<()> {
    let webview_window = app.get_webview_window("Main").unwrap();
    webview_window.hide().unwrap();
    webview_window.set_decorations(false).unwrap();
    
    let target_monitor = gtk::gdk::Display::default().unwrap()
        .monitor(0).unwrap();
    
    let window = gtk::ApplicationWindow::new(&webview_window.gtk_window().unwrap().application().unwrap());
    
    window.set_app_paintable(true);
    window.set_decorated(false);
    window.stick();
    
    webview_window.gtk_window().unwrap().remove(&webview_window.default_vbox().unwrap());
    window.add(&webview_window.default_vbox().unwrap());

    window.init_layer_shell();

    window.auto_exclusive_zone_enable();

    window.set_layer(Layer::Top);
    window.set_anchor(Edge::Left, true);
    window.set_width_request(50);
    window.set_height_request(target_monitor.geometry().height());

    window.show_all();

    window.set_monitor(&target_monitor);

    Ok(())
}
```

**Comment by LucaCoduriV** [comment]
Oh wow, so it means that by using the official crate I can use layer shells ?

I remember that I already tried what you did a long time ago and it wasn't working, but the might changed something since.

**Comment by andrewbaxter** [comment]
Oh no way, I was just looking for a notification daemon and found https://github.com/LucaCoduriV/Notiflut-Land and I thought I knew the author from somewhere... 😅

**Comment by LucaCoduriV** [comment]
> Oh no way, I was just looking for a notification daemon and found https://github.com/LucaCoduriV/Notiflut-Land and I thought I knew the author from somewhere... 😅 

😂
Yeah I tried some things but I stopped.
My code wasn't good and it was hard and annoying to use flutter with rust. But I really want to redo it better when I have time and using only rust this time. That's the main reason of why I want to be able to use layer shell with tauri.

**Comment by darman96** [comment]
Cool to see that there are more people interested in doing Desktop Widgets/Shells with web tech.
I'm currently building something similar in C++

**Comment by darman96** [comment]
If you are interested, here's my progress so far:
https://github.com/darman96/web-shell
https://github.com/darman96/web-shell-widgets

In my widgets repo I setup vite so that the widgets are build as individual single file pages by having a separate vite config per widget. When I run the dev server I have a main component to emulate a desktop where I can place all the widgets.

Communication with the main app will be handled by a WebProcessExtension which communicates with the main app via IPC.

**Comment by andrewbaxter** [comment]
Oh nice! Need some screenshots :D

**Comment by darman96** [comment]
Here's my current WIP Launcher widget:
![Image](https://github.com/user-attachments/assets/46b96f8c-2df9-4d2d-a02e-8d54c9e4962d)

**Comment by darman96** [comment]
Me and copilot have been hard at work tonight 😁 
I started over in rust

https://github.com/user-attachments/assets/ebf22201-dce2-4781-b9ca-798022b96da7

Dynamic resizing layer shell, the size of the react widget is applied to the window
